### PR TITLE
backend,frontend: implement AOPP workflow

### PR DESCRIPTION
**Pull request by benma**
More unit tests to follow. This is a start, but I tried to be as complete as possible. If the concept makes sense, ideally we merge this and iterate with improving the workflow, UI and edge cases.

One can test by going to https://testing.aopp.group/ and clicking the link after installing the app (on macOS: dragging the app to the Applications folder).

You can also copy the link and launch the app on the command line with the aopp URI as the argument, e.g.:

./BitBox 'aopp:?v=0&msg=I+confirm+that+this+Bitcoin+%28BTC%29+address+is+controlled+by+David+Lopez%2C+Poststrasse+22%2C+Zug%2C+Switzerland.+Unique+Identifier%3A+b83762dfbd8696a&asset=btc&format=any&callback=https%3A%2F%2Ftesting.21analytics.ch%2Fproofs%2F3730c437-4c4a-4364-8e7c-1b5a84b039f2'

@thisconnect to iteratively develop with `make webdev`, I used this patch in the backend to invoke the AOPP workflow a few seconds after the backend starts: https://github.com/benma/bitbox-wallet-app/commit/0529615c76c1cb6f2845ef6b9fe4a6aeee171269

### CODE REVIEW

File: backend/aopp.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	backend.aopp.State = aoppStateChoosingAccount
+	backend.notifyAOPP()
+}
+
+// handleAOPP handles an AOPP (Address Ownership Proof Protocol) request. See https://aopp.group/.
+func (backend *Backend) handleAOPP(uri *url.URL) {
+	defer backend.accountsAndKeystoreLock.Lock()()
+
+	log := backend.log.WithField("aopp-uri", uri.String())
+
```
Comment (x1ddos) [inline backend/aopp.go:154]: lots of whitespace :space_invader:

File: backend/aopp.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	if q.Get("v") != "0" {
+		log.Error("Can only handle version 0 aopp URIs")
+		backend.aoppSetError(errAoppVersion)
+		return
+	}
+
+	callback := q.Get("callback")
+	if callback == "" {
+		log.Error("callback param missing")
+		return
```
Comment (x1ddos) [inline backend/aopp.go:168]: don't you want to set error to `errAoppInvalidRequest` here, before returning?

File: backend/aopp.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	}
+
+	backend.aopp.Accounts = accounts
+	backend.aopp.State = aoppStateChoosingAccount
+	backend.notifyAOPP()
+}
+
+// handleAOPP handles an AOPP (Address Ownership Proof Protocol) request. See https://aopp.group/.
+func (backend *Backend) handleAOPP(uri *url.URL) {
```
Comment (x1ddos) [inline backend/aopp.go:150]: why pointer?

File: backend/aopp.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+	defer backend.accountsAndKeystoreLock.Lock()()
+
+	log := backend.log.WithField("aopp-uri", uri.String())
+
+	q := uri.Query()
```
Comment (x1ddos) [inline backend/aopp.go:155]: since `uri` arg is a pointer, this can lead to null pointer dereference.

File: backend/aopp.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+		return
+	}
+	backend.aopp.CallbackHost = callbackURL.Host
+	backend.aopp.callback = callback
+
+	coinCode, ok := map[string]coinpkg.Code{
+		"btc": coinpkg.CodeBTC,
+		// TODO: add support for ETH
+		// "eth": coinpkg.CodeETH,
+	}[strings.ToLower(q.Get("asset"))]
```
Comment (x1ddos) [inline backend/aopp.go:183]: would be much nicer as a package-global map. it's very hidden here.

File: backend/aopp.go [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+		return
+	}
+	backend.aopp.message = msg
+
+	backend.aopp.State = aoppStateAwaitingKeystore
+	if backend.keystore == nil {
+		backend.notifyAOPP()
+		return
+	}
+	backend.aoppKeystoreRegistered()
```
Comment (x1ddos) [inline backend/aopp.go:204]: I'm confused. aoppKeystoreRegistered doc comment says:

> aoppKeystoreRegistered must be called after a keystore is available

but it is called here, in `handleAOPP`. would've expected `aoppKeystoreRegistered` to be called elsewhere like here: https://github.com/digitalbitbox/bitbox-wallet-app/blob/1445822/backend/backend.go#L928

File: backend/aopp.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+// when calling this function.
+func (backend *Backend) aoppSetError(err ErrorCode) {
+	backend.aopp.State = aoppStateError
+	backend.aopp.ErrorCode = err
+	backend.notifyAOPP()
+}
+
+// aoppKeystoreRegistered must be called after a keystore is available, to display a list of
+// accounts to choose from. `accountsAndKeystoreLock` must be held when calling this function.
+func (backend *Backend) aoppKeystoreRegistered() {
```
Comment (x1ddos) [inline backend/aopp.go:117]: please add a comment that this is also called right away, "internally" by handleAOPP if a keystore is already available at the time AOPP is initiated.

File: backend/aopp.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+	defer backend.accountsAndKeystoreLock.Lock()()
+	if backend.aopp.State != aoppStateChoosingAccount {
+		return
+	}
+
+	backend.aopp.State = aoppStateSyncing
+	backend.notifyAOPP()
+
+	log := backend.log.WithField("accountCode", code)
+	account := func() accounts.Interface {
```
Comment (x1ddos) [inline backend/aopp.go:219]: don't really need a closure func. it gets shorter:

```go
var account accounts.Interface
for _, acct := range backend.accounts {
    if acct.Config().Code == code {
      account = acct
    }
}
// account is nil if not found
```

File: backend/backend.go [hunk 9]
```diff
@@ -689,6 +693,15 @@ func (backend *Backend) Banners() *banners.Banners {
...
-	// TODO: handle
-	backend.log.Debugf("Handle URI: %s", uri)
+	u, err := url.Parse(uri)
+	if err != nil {
+		backend.log.WithError(err).Warningf("Handling URI failed: %s", uri)
+		return
+	}
+	switch u.Scheme {
+	case "aopp":
+		backend.handleAOPP(u)
```
Comment (x1ddos) [inline backend/backend.go:703]: I would pass a copy here:

```go
backend.handleAOPP(*u)
```

avoid worring about null pointer dereference in handleAOPP. https://play.golang.org/p/d9YQrWZJeYb

**Comment by benma** [comment]
Added one unit test exercising only the success path. More to follow.

**Review by x1ddos — COMMENTED** [review]
Looks good. Haven't looked at the frontend yet but I assume @thisconnect will.

File: backend/aopp.go (hunk 6 above)
Comment (x1ddos) [reply backend/aopp.go:204]: never mind this comment. it was before I went through the rest.

File: backend/aopp.go (hunk 4 above)
Comment (benma) [reply backend/aopp.go:155]: isn't that just how Go works? half the stuff is passed by pointer to avoid copies :(

File: backend/backend.go (hunk 9 above)
Comment (benma) [reply backend/backend.go:703]: On the other hand it copies lots of bytes for no reason?

File: backend/aopp.go (hunk 3 above)
Comment (benma) [reply backend/aopp.go:150]: cause this is what url.Parse() returns :innocent:

File: backend/errors.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	// errAoppUnsupportedAsset is returned when an AOPP request is for an asset we don't support AOPP
+	// for.
+	errAoppUnsupportedAsset ErrorCode = "aoppUnsupportedAsset"
+	// errAoppVersion is returned when we cannot handle an AOPP request because we don't support the
+	// request version.
+	errAoppVersion ErrorCode = "aoppVersion"
+	// errAoppNoAccounts is returned when there are no available accounts to choose from.
+	errAoppNoAccounts ErrorCode = "aoppNoAccounts"
+	// errAoppCannotSign is returned when the connected keystore does not support signing messages.
+	errAoppCannotSign ErrorCode = "aoppCannotSign"
```
[the remark this answers is not in the document]
Comment (benma) [reply backend/errors.go]: :heavy_check_mark:

File: backend/errors.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+const (
+	// ErrAccountAlreadyExists is returned if an account is being added which already exists.
+	ErrAccountAlreadyExists ErrorCode = "accountAlreadyExists"
+	// ErrAccountLimitReached is returned when adding an account if no more accounts can be added.
+	ErrAccountLimitReached ErrorCode = "accountLimitReached"
+
+	// errAoppUnsupportedAsset is returned when an AOPP request is for an asset we don't support AOPP
+	// for.
+	errAoppUnsupportedAsset ErrorCode = "aoppUnsupportedAsset"
```
[the remark this answers is not in the document]
Comment (benma) [reply backend/errors.go]: :heavy_check_mark:

File: backend/aopp.go (hunk 1 above)
Comment (benma) [reply backend/aopp.go]: :heavy_check_mark:

File: backend/aopp.go (hunk 2 above)
Comment (benma) [reply backend/aopp.go:175]: :heavy_check_mark:

File: backend/aopp.go (hunk 3 above)
Comment (benma) [reply backend/aopp.go]: Passing by value now.

File: backend/aopp.go (hunk 4 above)
Comment (benma) [reply backend/aopp.go:162]: Passing by value now.

File: backend/aopp.go (hunk 5 above)
Comment (benma) [reply backend/aopp.go]: :heavy_check_mark:

File: backend/aopp.go (hunk 7 above)
Comment (benma) [reply backend/aopp.go:126]: :heavy_check_mark:

File: backend/aopp.go (hunk 8 above)
Comment (benma) [reply backend/aopp.go]: :heavy_check_mark:

File: backend/backend.go (hunk 9 above)
Comment (benma) [reply backend/backend.go]: Done

File: backend/aopp.go (hunk 4 above)
Comment (x1ddos) [reply backend/aopp.go:162]: I wouldn't worry about memory optimization here. There are lots of much bigger allocations in the app. But if you do worry, here's an article: https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963

Takeaway:

> Using a copy of the struct instead of a pointer here is 8 times faster.

**Review by x1ddos — APPROVED** [review]
LGTM. passing on to @thisconnect.

File: frontends/web/src/components/aopp/aopp.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+                    <div>
+                        <p>Address: {aopp.address}</p>
+                        <p>Please confirm on your BitBox.</p>
+                    </div>
+                );
+            case 'success':
+                return (
+                    <div>
+                        <p>Successfully delivered a fresh address to the third party.</p>
+                        <Button primary onclick={backendAPI.aoppCancel}>Dismiss</Button>
```
Comment (thisconnect) [inline frontends/web/src/components/aopp/aopp.tsx:113]: There are a few strings to add to i18n and translate

File: frontends/web/src/api/backend.ts [hunk 13]
```diff
@@ -49,3 +49,26 @@ export const renameAccount = (accountCode: string, name: string): Promise<ISucce
...
+    callbackHost: string;
+}
+
+export const aoppCancel = (): Promise<null> => {
+    return apiPost('aopp/cancel');
+};
+
+export const aoppChooseAccount = (accountCode: AccountCode): Promise<null> => {
+    return apiPost('aopp/choose-account', { accountCode });
+};
```
Comment (thisconnect) [inline frontends/web/src/api/backend.ts:74]: I'd move this to `api/aopp.ts` but can do later

**Review by thisconnect — APPROVED** [review]
LGTM, added some feedback but I can pick it up and make a commit ontop of this branch.

File: util/observable/action/action.go [hunk 14]
```diff
@@ -18,6 +18,10 @@ package action
 type Action string
 
 const (
+	// Notify pushes information to the listener. The listener handles the notification depending on
+	// the subject.
+	Notify Action = "notify"
```
[the remark this answers is not in the document]
Comment (benma) [reply util/observable/action/action.go:23]: dang missed it again lol. No it is not, will remove.

File: frontends/web/src/components/aopp/aopp.tsx [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+                );
+        }
+    }
+}
+
+const subscribeHOC = subscribe<SubscribedProps, AoppProps & TranslateProps>(
+    { aopp: 'aopp' },
+    false,
+    false,
+)(Aopp);
```
[the remark this answers is not in the document]
Comment (benma) [reply frontends/web/src/components/aopp/aopp.tsx:124]: I have really a hard time doing this, because in my mind, the subscribe decorator is perfect. It does the apiGet on first load, it gets the subsequent notifications, it does not clutter the component with multiple calls to do the same in didMount... what is not to like about it?

File: frontends/web/src/components/aopp/aopp.tsx (hunk 12 above)
Comment (benma) [reply frontends/web/src/components/aopp/aopp.tsx:113]: i left them literal because this is a prototype, maybe you will change everything anyway ;) i just put into i18n what i knew was gonna be needed for sure.

File: frontends/web/src/api/backend.ts (hunk 13 above)
Comment (benma) [reply frontends/web/src/api/backend.ts:74]: Done

File: util/observable/action/action.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (benma) [reply util/observable/action/action.go]: Removed

File: frontends/web/src/components/aopp/aopp.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/src/components/aopp/aopp.tsx:124]: lets save this discussing for another day.

File: frontends/web/src/components/aopp/aopp.tsx [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+import { Component, h, RenderableProps } from 'preact';
+import { AccountCode } from '../../api/account';
+import * as aoppAPI from '../../api/aopp';
+import { subscribe } from '../../decorators/subscribe';
+import { translate, TranslateProps } from '../../decorators/translate';
+import { Button, Select } from '../forms';
+
+interface State {
+}
```
Comment (thisconnect) [inline frontends/web/src/components/aopp/aopp.tsx:25]: missing types

File: frontends/web/src/components/aopp/aopp.tsx [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+type Props = SubscribedProps & AoppProps & TranslateProps;
+
+class Aopp extends Component<Props, State> {
+    public readonly state: State = {
+        aopp: {
+            state: 'inactive',
+            accounts: [],
+            errorCode: '',
+            address: '',
```
Comment (thisconnect) [inline frontends/web/src/components/aopp/aopp.tsx:42]: there is also callbackHost

Screen Shot 2021-06-24 at 9 49 36 AM

**Review by thisconnect — APPROVED** [review]

File: frontends/web/src/components/aopp/aopp.tsx (hunk 16 above)
Comment (benma) [reply frontends/web/src/components/aopp/aopp.tsx:25]: there are no State vars.

File: frontends/web/src/components/aopp/aopp.tsx (hunk 17 above)
Comment (benma) [reply frontends/web/src/components/aopp/aopp.tsx:42]: fixed
